dask/array/core.py

```diff
@@ -1938,14 +1982,49 @@ def normalize_chunks(chunks, shape=None):
             raise ValueError("Input array has %d dimensions but the supplied "
                              "chunks has only %d dimensions" %
                              (len(shape), len(chunks)))
-        if not all(c == s or (math.isnan(c) and math.isnan(s))
+        if not all(c == s or (math.isnan(c) or math.isnan(s))
                    for c, s in zip(map(sum, chunks), shape)):
             raise ValueError("Chunks do not add up to shape. "
                              "Got chunks=%s, shape=%s" % (chunks, shape))
 
     return tuple(tuple(int(x) if not math.isnan(x) else x for x in c) for c in chunks)
 
 
+def auto_chunks(chunks, shape, limit, itemsize):
+    """ Determine automatic chunks
+
+    See also
+    --------
+    normalize_chunks: for full docstring and parameters
+    """
+    assert limit and itemsize
+    limit = limit // itemsize
+    chunks = list(chunks)
+
+    autos = [i for i, c in enumerate(chunks) if c == 'auto']
+    if not autos:
+        return tuple(chunks)
+
+    largest_block = np.prod([cs if isinstance(cs, Number) else max(cs)
+                             for cs in chunks if cs != 'auto'])
+    size = (limit / largest_block) ** (1 / len(autos))
```

Comment:
Interesting. This is different from the algorithm I had in mind:
- Your approach sets all `'auto'` chunked dimensions to use the same chunk size. This results in "square" chunks, with small dimensions having a single chunk.
- I was thinking of multiplying original chunk-sizes along `'auto'` dimensions by a constant factor. This results in chunks with the same aspect ratio as the original chunks.

I think my approach would be more memory friendly, especially in the case of streaming processing, though arguably the resulting chunks are less intuitive. I'm thinking particularly of use cases like a large dask array, where you want `array.rehcunk('auto')` to rechunk to make scalar array operations fast, e.g., a stack of 1000 images of size 1000x1000 with 8 bytes/element. The current logic would result in chunks like (250, 250, 250) whereas my proposal would give (12, 1000, 1000).